### PR TITLE
feat(train): Support gradient checkpointing for Conformer & Transformer

### PR DESCRIPTION
**Pull request by xingchensong**
# TODO

- [ ] benchmark

### CODE REVIEW

**Comment by robin1001** [comment]
Reduce the memory in training, and the cost is slower training speed?

**Comment by xingchensong** [comment]
> Reduce the memory in training, and the cost is slower training speed?

yes, "fit 10x larger neural nets into memory at the cost of an additional 20% computation time"

**Comment by xingchensong** [comment]
I think it's possible to fintune whisper-large-v3 (full parameter) on 24 * 2080ti, by Deepspeed + Grad Checkpointing + Flash Attention.

**Comment by xingchensong** [comment]
img_v3_025l_f327d50d-2558-4e5a-aea3-24275575b55g

**Comment by xingchensong** [comment]
- conformer.wenet.2.2.1: ddp
- conformer.latest.373109c.grad_ckpt： ddp + grad_ckpt
- conformer.latest.373109c.grad_ckpt.ds: deepspeed stage1 + grad_ckpt

image

1. cv_loss is almost identical to wenet v2.2.1
2. training with grad_ckpt has 15%~20% additional computation time compared to v2.2.1

**Comment by Mddct** [comment]
fp32？

**Comment by xingchensong** [comment]
> fp32？

yes ！！!

**Comment by Mddct** [comment]
amazing again！！！

**Comment by xingchensong** [comment]
TBD: `gradient_checkpointing` 应该在yaml中配置还是在train args中配置？目前是在yaml中配置的，不在args中配置，是考虑到目前只有transformer/conformer支持了grad_ckpt，其他模型都没支持，用户使用 `squeezeformer` 等模型时如果加了参数 `--gradient_checkpoiting` ，实际并不起作用（因为squeezeformer的forward里没实现这个）

**Comment by xingchensong** [comment]
pass unit test (gradient is equal), ready for a final review

**Comment by Mddct** [comment]
> wait, I just find that `find_unused_param=False` actually works in conjunction with `ctc_weight=0.0`. Perhaps we could simply eliminate the argument and retain `find_unused_param=False` as our default configuration.

与模型有关的 可能放yaml里会好些， 也不想影响其他模块

**Comment by xingchensong** [comment]
> What is gradient ckpt？
一句话总结: 2080ti 全量参数微调whisper不再是梦😜（具体技术细节可见开头的blog 链接）

**Comment by xingchensong** [comment]
pass JIT ! please review this PR, @robin1001 @Mddct

**Review by Mddct — APPROVED** [review]
